File: 35.search-insert-position.java
/*
 * @lc app=leetcode id=35 lang=java
 *
 * [35] Search Insert Position
 */

// @lc code=start
class Solution {
    public int searchInsert(int[] nums, int target) {
        return binarySearch(nums, target, 0, nums.length - 1);

    }

    int binarySearch(int[] arr, int tar, int start, int end) {
        if (start > end)
            return start;
        int mid = Math.round((start + end) / 2);
        if (arr[mid] == tar)
            return mid;
        if (arr[mid] > tar)
            return binarySearch(arr, tar, start, mid - 1);
        if (arr[mid] < tar)
            return binarySearch(arr, tar, mid + 1, end);
        return 0;
    }
}
// @lc code=end
// 62/62 cases passed (0 ms)
// Your runtime beats 100 % of java submissions
// Your memory usage beats 100 % of java submissions (39.1 MB)

File: 35.search-insert-position.js
/*
 * @lc app=leetcode id=35 lang=javascript
 *
 * [35] Search Insert Position
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function(nums, target) {
  return binarySearch(nums, target, 0, nums.length - 1);

  function binarySearch(arr, target, start, end) {
    if (start > end) return start;
    const mid = Math.floor((start + end) / 2);
    if (arr[mid] === target) return mid;
    if (arr[mid] > target) return binarySearch(arr, target, start, mid - 1);
    if (arr[mid] < target) return binarySearch(arr, target, mid + 1, end);
  }
};
// @lc code=end
// 62/62 cases passed (52 ms)
// Your runtime beats 81.52 % of javascript submissions
// Your memory usage beats 62.5 % of javascript submissions (33.8 MB)
